### PR TITLE
add avro schema serializer and deserializer

### PR DESCRIPTION
**Pull request by humkum**
add avro schema serializer and deserializer for rocketmq schema registry

### CODE REVIEW

File: client/src/main/java/org/apache/rocketmq/schema/registry/client/serializer/AbstractAvroDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(bais, null);
+
+        ByteBuffer buffer = ByteBuffer.allocate(16);
+        try {
+            decoder.readBytes(buffer);
+        } catch (Exception e) {
+            log.error("read bytes error: ", e);
+        }
+
+        long id = buffer.getLong();
```
Comment (ferrirW) [inline client/src/main/java/org/apache/rocketmq/schema/registry/client/serializer/AbstractAvroDeserializer.java:73]: id + version

File: client/src/main/java/org/apache/rocketmq/schema/registry/client/serializer/AbstractAvroSerializer.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+        try {
+
+        } catch (Exception e) {
+            throw new SerializationException("get schema by subject failed", e);
+        }
+
+        try {
+            SchemaRecordDto schemaRecordDto = getSchemaBySubject(subject);
+            long schemaId = schemaRecordDto.getSchemaId();
```
Comment (ferrirW) [inline client/src/main/java/org/apache/rocketmq/schema/registry/client/serializer/AbstractAvroSerializer.java:60]: in our design, we use id + version to place a schema record, so you should write version in header too

**Review by ferrirW — APPROVED** [review]

**Comment by MatrixHB** [comment]
I have an idea about design of schema serializer/deserializer

What's exposed to users is `SchemaWrapper`,  which includes `Map<String/*SchemaType*/, Serializer>` and `Map<String/*SchemaType*/, Deserializer>`(such as <AVRO,  AvroSerializer>, or use Factory...).  Then the user doesn't need to know the schema type (avro/json/thrift....) and just leave the serializer work to `SchemaWrapper`. 
In `SchemaWrapper`, after getting schema from registry, we can know the type of schema, then we can decide what kind of serializer to use.

The pseudo code as follows.
```
public class SchemaWrapper {

    private final SchemaRegistryClient schemaRegistryClient;

    private Map<String/*schema type*/, Serializer> serializerMap;
    
    <T> byte[] serialize(T data, String topic) throws Exception{
        GetSchemaResponse response = schemaRegistryClient.getSchemaBySubject(topic);
        Serializer serializer = serializerMap.get(response.getType);
        return serializer.serialize(data, response.getIdl());
    }

    
    <T> T deserialize(byte[] data, String topic) {
        
    }
}

```
@humkum @ferrirW  What do you think about this design?

**Comment by humkum** [comment]
> I have an idea about design of schema serializer/deserializer
> 
> What's exposed to users is `SchemaWrapper`, which includes `Map<String/*SchemaType*/, Serializer>` and `Map<String/*SchemaType*/, Deserializer>`(such as <AVRO, AvroSerializer>, or use Factory...). Then the user doesn't need to know the schema type (avro/json/thrift....) and just leave the serializer work to `SchemaWrapper`. In `SchemaWrapper`, after getting schema from registry, we can know the type of schema, then we can decide what kind of serializer to use.
> 
> The pseudo code as follows.
> 
> ```
> public class SchemaWrapper {
> 
>     private final SchemaRegistryClient schemaRegistryClient;
> 
>     private Map<String/*schema type*/, Serializer> serializerMap;
>     
>     <T> byte[] serialize(T data, String topic) throws Exception{
>         GetSchemaResponse response = schemaRegistryClient.getSchemaBySubject(topic);
>         Serializer serializer = serializerMap.get(response.getType);
>         return serializer.serialize(data, response.getIdl());
>     }
> 
>     
>     <T> T deserialize(byte[] data, String topic) {
>         
>     }
> }
> ```
> 
> @humkum @ferrirW What do you think about this design?

good idea!

**Comment by ferrirW** [comment]
> I have an idea about design of schema serializer/deserializer
> 
> What's exposed to users is `SchemaWrapper`, which includes `Map<String/*SchemaType*/, Serializer>` and `Map<String/*SchemaType*/, Deserializer>`(such as <AVRO, AvroSerializer>, or use Factory...). Then the user doesn't need to know the schema type (avro/json/thrift....) and just leave the serializer work to `SchemaWrapper`. In `SchemaWrapper`, after getting schema from registry, we can know the type of schema, then we can decide what kind of serializer to use.

Yes, that's why StorageInfo exist in SchemaInfo, we can use this inforamtion to manage user's ser/de process and data format.

But now server side is not ready yet. And in my opinion, exposes the type of Avro/PB/JSON to user is more easier understand and change.

How about that as a follow-up improvement? @MatrixHB

**Comment by MatrixHB** [comment]
> > I have an idea about design of schema serializer/deserializer
> > What's exposed to users is `SchemaWrapper`, which includes `Map<String/*SchemaType*/, Serializer>` and `Map<String/*SchemaType*/, Deserializer>`(such as <AVRO, AvroSerializer>, or use Factory...). Then the user doesn't need to know the schema type (avro/json/thrift....) and just leave the serializer work to `SchemaWrapper`. In `SchemaWrapper`, after getting schema from registry, we can know the type of schema, then we can decide what kind of serializer to use.
> 
> Yes, that's why StorageInfo exist in SchemaInfo, we can use this inforamtion to manage user's ser/de process and data format.
> 
> But now server side is not ready yet. And in my opinion, exposes the type of Avro/PB/JSON to user is more easier understand and change.
> 
> How about that as a follow-up improvement? @MatrixHB

 @ferrirW  I agree that it can be a follow-up improvement. 
However, now the `SchemaRegistryClient` is injected into AvroSerializer/AvroDeserializer. If we want to use StorageInfo or SchemaType to decide ser/de process, we need to inject `SchemaRegistryClient` and the `getSchemaBySubject` logic into SchemaWapper but not Ser/De.
Will it be difficult to change when doing follow-up improvement?

**Pull request closed without merging**